design/backup-performance-improvements.md

```diff
@@ -0,0 +1,357 @@
+- Per-backup worker pool advantages:
+  - Lower memory consumption than permanent worker pool, but the total memory used by a worker blocked on input will be pretty low, so if we're talking only 10-20 workers, the impact will be minimal.
+
+## Compatibility
+
+### Example IBA implementation for BIA plugins which return additional items
+
+Included below is an example of what might be required for a BIA  plugin which returns additional items.
+The code is taken from the internal velero `pod_action.go` which identifies the items required for a given pod.
+
+In this particular case, the only function of pod_action is to return additional items, so we can really just convert this plugin to an IBA plugin. If there were other actions, such as modifying the pod content on backup, then we would still need the pod action, and the related items vs. content manipulation functions would need to be separated.
+
+```go
+// PodAction implements ItemBlockAction.
+type PodAction struct {
+	log logrus.FieldLogger
+}
+
+// NewPodAction creates a new ItemAction for pods.
+func NewPodAction(logger logrus.FieldLogger) *PodAction {
+	return &PodAction{log: logger}
+}
+
+// AppliesTo returns a ResourceSelector that applies only to pods.
+func (a *PodAction) AppliesTo() (velero.ResourceSelector, error) {
+	return velero.ResourceSelector{
+		IncludedResources: []string{"pods"},
+	}, nil
+}
+
+// GetRelatedItems scans the pod's spec.volumes for persistentVolumeClaim volumes and returns a
+// ResourceIdentifier list containing references to all of the persistentVolumeClaim volumes used by
+// the pod. This ensures that when a pod is backed up, all referenced PVCs are backed up too.
+func (a *PodAction) GetRelatedItems(item runtime.Unstructured, backup *v1.Backup) (runtime.Unstructured, []velero.ResourceIdentifier, error) {
+	pod := new(corev1api.Pod)
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(item.UnstructuredContent(), pod); err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	var relatedItems []velero.ResourceIdentifier
+	if pod.Spec.PriorityClassName != "" {
+		a.log.Infof("Adding priorityclass %s to relatedItems", pod.Spec.PriorityClassName)
+		relatedItems = append(relatedItems, velero.ResourceIdentifier{
+			GroupResource: kuberesource.PriorityClasses,
+			Name:          pod.Spec.PriorityClassName,
+		})
+	}
+
+	if len(pod.Spec.Volumes) == 0 {
+		a.log.Info("pod has no volumes")
+		return relatedItems, nil
+	}
+
+	for _, volume := range pod.Spec.Volumes {
```

Comment:
For such scenarios 
Do we want to support the other filtering mechanisms such as label selector, include/exclude resource group type 
Trying to understand what will take precedence here.

Will being a related item force backup of all dependant resources, disregarding the filters.
Or will filters be enforced post related resources discovery and hence might eventually fail the itemblock in some way?

I understand that in current BIA flow, in some cases we do forcefully backup stuff.

Reply:
@anshulahuja98 The above is the current logic taken from pod_action (where we add PVCs for each pod as AdditionalItems). In any case include/exclude/labels for backup isn't something that the plugins do, that's handled in the velero logic. I believe includes/excludes are *already* considered, and additional items excluded via resource type are already filtered (so that won't change here), but I don't think labels are considered there -- although I'm not certain -- but again, that's something that shouldn't change here. This design is about sequencing and threading of items backed up, not about modifying the set of backed up items. Any changes we want to make to that behavior would be out-of-scope for this design.

Reply:
just want to make sure we don't impact existing behaviour in any way after this change. 
Feel free to resolve this since you are certain that won't happen. 

You can see if you want to call that out in the design in somewhere so we don't miss.

Any change in current velero behaviour if needed we can address separately

Reply:
@anshulahuja98 I'm not really sure I can think of anything we'd want to change about how plugins call for required additional items. We already have an issue about what to do about label selectors and additional items, so that has a place to be addressed (I think for the moment, we're not planning on changing anything, if I'm remembering the issue conversation correctly), and the includes/excludes filtering is already done on additional items, which won't change in this design.

Reply:
Fair enough
Can you point me to the issue where discussed?
If handy

Reply:
@anshulahuja98 https://github.com/vmware-tanzu/velero/issues/7750 is the issue I was referring to.